src/hotspot/share/oops/methodData.cpp

```diff
@@ -1821,3 +1821,12 @@ void MethodData::clean_weak_method_links() {
   clean_extra_data(&cl);
   verify_extra_data_clean(&cl);
 }
+
+void MethodData::deallocate_contents(ClassLoaderData* loader_data) {
+  assert(_extra_data_lock == NULL, "memory leak");
```

Comment:
I don't know why you didn't make deallocate_contents call release_C_heap_structures, like the other metadata deallocate_contents functions do?

Reply:
I see why.  InstanceKlass release_C_heap_structures is called for class unloading so frees everything in its tree, and deallocate_contents does the same thing except for the constant pool, because the ConstantPool deallocate_contents calls its own release_C_heap_structures.  Methods are attached to InstanceKlass so let InstanceKlass call release_C_heap_structures for them, except for this case that was added for MethodData here.